src/charm.py

```diff
@@ -2,22 +2,208 @@
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-"""Charmed operator for the 5G NSSF service."""
+"""Charmed operator for the SDCORE NSSF service."""
 
 import logging
+from ipaddress import IPv4Address
+from subprocess import check_output
 
-from ops.charm import CharmBase
+from charms.observability_libs.v1.kubernetes_service_patch import (  # type: ignore[import]
+    KubernetesServicePatch,
+)
+from charms.prometheus_k8s.v0.prometheus_scrape import (  # type: ignore[import]
+    MetricsEndpointProvider,
+)
+from charms.sdcore_nrf.v0.fiveg_nrf import NRFRequires  # type: ignore[import]
+from jinja2 import Environment, FileSystemLoader
+from lightkube.models.core_v1 import ServicePort
+from ops.charm import CharmBase, EventBase
 from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import Layer
 
 logger = logging.getLogger(__name__)
 
+PROMETHEUS_PORT = 9089
+SBI_PORT = 29531
+CONFIG_DIR = "/free5gc/config"
+CONFIG_FILE_NAME = "nssfcfg.conf"
+CONFIG_TEMPLATE_DIR = "src/templates/"
+CONFIG_TEMPLATE_NAME = "nssfcfg.conf.j2"
+
 
 class NSSFOperatorCharm(CharmBase):
-    """Charm the service."""
+    """Main class to describe juju event handling for the 5G NSSF operator."""
 
     def __init__(self, *args):
         super().__init__(*args)
+        self._nssf_container_name = self._nssf_service_name = "nssf"
+        self._nssf_container = self.unit.get_container(self._nssf_container_name)
+        self._nrf_requires = NRFRequires(charm=self, relation_name="fiveg_nrf")
+        self._nssf_metrics_endpoint = MetricsEndpointProvider(self)
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[
+                ServicePort(name="prometheus-exporter", port=PROMETHEUS_PORT),
+                ServicePort(name="sbi", port=SBI_PORT),
+            ],
+        )
+
+        self.framework.observe(self.on.nssf_pebble_ready, self._on_nssf_pebble_ready)
+        self.framework.observe(self._nrf_requires.on.nrf_available, self._on_nssf_pebble_ready)
+
+    def _on_nssf_pebble_ready(
+        self,
+        event: EventBase,
+    ) -> None:
+        """Handle pebble ready event for NSSF container.
+
+        Args:
+            event (EventBase): Juju event
+        """
+        if not self._nssf_container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to start")
+            event.defer()
+            return
+        if not self._relation_created("fiveg_nrf"):
+            self.unit.status = BlockedStatus("Waiting for fiveg_nrf relation")
+            return
+        if not self._nrf_data_is_available:
+            self.unit.status = WaitingStatus("Waiting for NRF data to be available")
+            event.defer()
+            return
+        if not self._nssf_container.exists(path=CONFIG_DIR):
```

Comment:
Following [this comment](https://github.com/canonical/sdcore-amf-operator/pull/1#discussion_r1197067742) on the `amf` pr I think we can follow the same for all sdcore charms that need to push a config file and in case it does exist on the pod compare its content the one we are pushing and push if the content is different.

Reply:
I see in the tests you are testing that the new content of a config is pushed if the content is changed, I assume the intended behaviour is to push anytime, and if it was changed it will be pushed anyway, which means that we are pushing the config file even when there is no change. 
Am I right? is this the intention?

Reply:
The config is passed to the startup command, I think we need to restart the workload to get the changes applied.

Reply:
I added a restart logic similar to what is in the UDR charm.